crates/qasm3/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+        Some(gates) => gates
+            .into_iter()
+            .map(|gate| (gate.name().to_owned(), gate))
+            .collect(),
+        None => module
+            .getattr(STDGATES_INC_CUSTOM_GATES_ATTR)?
+            .iter()?
+            .map(|obj| {
+                let gate = obj?.extract::<circuit::PyGate>()?;
+                Ok((gate.name().to_owned(), gate))
+            })
+            .collect::<PyResult<HashMap<_, _>>>()?,
+    };
+    crate::build::convert_asg(py, result.program(), result.symbol_table(), gates)
+}
+
+/// Load an OpenQASM 3 program from a source file into a :class:`.QuantumCircuit`.
+///
+/// Args:
+///     pathlike_or_filelike (str | os.PathLike | io.TextIOBase): the program source.  This can
+///         either be given as a filepath, or an open text stream object.  If the stream is already
+///         opened it is consumed in Python space, whereas filenames are opened and consumed in
+///         Rust space; there might be slightly different performance characteristics, depending on
+///         your system and how the streams are buffered by default.
+///     custom_gates (Iterable[CustomGate]): Python constructors to use for particular named gates.
+///         If not supplied, Qiskit will use its own standard-library constructors for gates
+///         defined in the OpenQASM 3.0 standard-library file ``stdgates.inc``.
+///     include_path (Iterable[str]): the path to search when resolving ``include`` statements.
+///         If not given, Qiskit will arrange for this to point to a location containing
+///         ``stdgates.inc`` only.  Paths are tried in the sequence order.
+///
+/// Returns:
+///     :class:`.QuantumCircuit`: the constructed circuit object.
+///
+/// Raises:
+///     :class:`.QASM3ImporterError`: if an error occurred during parsing or semantic analysis.
+///         In the case of a parsing error, most of the error messages are printed to the terminal
+///         and formatted, for better legibility.
+#[pyfunction]
+#[pyo3(
+    pass_module,
+    signature = (pathlike_or_filelike, /, *, custom_gates=None, include_path=None),
+)]
+pub fn load(
+    module: &PyModule,
+    py: Python,
+    pathlike_or_filelike: &PyAny,
+    custom_gates: Option<Vec<circuit::PyGate>>,
+    include_path: Option<Vec<OsString>>,
+) -> PyResult<circuit::PyCircuit> {
+    let source =
+        if pathlike_or_filelike.is_instance(PyModule::import(py, "io")?.getattr("TextIOBase")?)? {
+            pathlike_or_filelike
+                .call_method0("read")?
+                .extract::<String>()?
+        } else {
+            let path = PyModule::import(py, "os")?
+                .getattr("fspath")?
+                .call1((pathlike_or_filelike,))?
+                .extract::<OsString>()?;
+            ::std::fs::read_to_string(&path).map_err(|err| {
+                QASM3ImporterError::new_err(format!("failed to read file '{:?}': {:?}", &path, err))
+            })?
+        };
```

Comment:
Looking at this I feel like we could just move the contents here to be pure python, I guess the only advantage of doing it in rust is if you can do `read_to_string()` which will avoid the python overhead and type conversion for file i/o.

Reply:
the main reason I did it in Rust was because I'd already written everything else in Rust - I wrote the `load` wrapper ages after I wrote the `loads` wrapper, and at the time I hadn't actually written the interface that wrapped them into `qiskit.qasm3`; I was just calling the functions from the Rust extension module directly.

I can pull `load` up to Python space completely, if you like and have it just call the Rust-space `loads`?

Reply:
This is fine, it doesn't really make a difference. I expect doing it in rust with a file path will be marginally more efficient because you avoid going through a Python string for the file contents.

It was more just when I was reading through and this and all the logic was done via python I was surprised.

Reply:
Avoiding conversion overhead for a string containing the OQ3 source makes sense to me. Handling the path name in Python is ok of course.